stream/stream_dvbv5.c

```diff
@@ -0,0 +1,687 @@
+    int r, ret;
+    struct dvb_v5_fe_parms *parms = NULL;
+    struct dvb_device *dvb;
+    struct dvb_dev_list *dvb_dev;
+    static dvb_priv_t cached_priv = {0};
+    dvb_priv_t *priv = NULL;
+    struct m_config_cache *opts_cache;
+    dvbv5_opts_t *opts;
+
+    opts_cache = m_config_cache_alloc(stream, stream->global, &stream_dvbv5_conf);
+    opts = opts_cache->opts;
+
+    if (opts->cfg_keep_tune && cached_priv.dvr_fd) {
+        stream->fill_buffer = dvbv5_streaming_read;
+        stream->close = dvbv5_close;
+        stream->control = dvbv5_stream_control;
+        stream->streaming = true;
+        stream->demuxer = "lavf";
+        stream->lavf_type = "mpegts";
+        stream->priv = talloc_memdup(stream, &cached_priv, sizeof(dvb_priv_t));
+        priv = stream->priv;
+        priv->opts_cache = opts_cache;
+        priv->opts = opts;
+        dvb = priv->dvb;
+        dvr_dev = priv->dvr_dev;
+        demux_dev = priv->demux_dev;
+        audio_fd = priv->audio_fd;
+        video_fd = priv->video_fd;
+        MP_INFO(stream, "Going cached!\n");
+
+
+        parms = dvb->fe_parms;
+        if (parse(stream, parms, &vpid, &apid, &sid)) {
+            dvb_dev_free(dvb);
+            return STREAM_ERROR;
+        }
+
+        uint32_t current_frequency, new_frequency;
+        dvb_fe_retrieve_parm(dvb->fe_parms, DTV_FREQUENCY, &current_frequency);
+        dvb_fe_retrieve_parm(parms, DTV_FREQUENCY, &new_frequency);
+
+        if (current_frequency == new_frequency) {
+            if (vpid >= 0) {
+                if (!video_fd) {
+                    MP_ERR(stream, "failed opening '%s'\n", demux_dev);
+                    dvb_dev_free(dvb);
+                    return STREAM_ERROR;
+                }
+
+                dvb_dev_set_bufsize(video_fd, DVB_BUF_SIZE);
+
+                if (dvb_dev_dmx_set_pesfilter(video_fd, vpid, DMX_PES_VIDEO,
+                    DMX_OUT_TS_TAP,
+                    64 * 1024) < 0) {
+                    dvb_dev_free(dvb);
+                    return STREAM_ERROR;
+                }
+            }
+
+            if (apid > 0) {
+                if (!audio_fd) {
+                    MP_ERR(stream, "failed opening '%s'\n", demux_dev);
+                    dvb_dev_free(dvb);
+                    return STREAM_ERROR;
+                }
+                if (dvb_dev_dmx_set_pesfilter(audio_fd, apid, DMX_PES_AUDIO,
+                        DMX_OUT_TS_TAP,
+                        64 * 1024) < 0) {
+                    dvb_dev_free(dvb);
+                    return STREAM_ERROR;
+                }
+            }
+            dvr_fd = dvb_dev_open(dvb, dvr_dev, O_RDONLY);
+            if (!dvr_fd) {
+                MP_ERR(stream, "failed opening '%s'\n", dvr_dev);
+                dvb_dev_free(dvb);
+                return STREAM_ERROR;
+            }
+
+            priv->dvr_fd = dvr_fd;
+            return STREAM_OK;
+        } else {
+            dvb_dev_free(priv->dvb);
```

Comment:
From the logic I see in `dvbv5-scan`, it should be possible to change frequency without closing the complete device. There's also more than frequency to it (polarization, DISEqC settings etc.) depending on the transport.